src/inc/til/small_vector.h

```diff
@@ -0,0 +1,845 @@
+#pragma warning(disable : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+
+namespace til
+{
+    // This class was adopted from std::span<>::iterator.
+    template<typename T>
+    struct small_vector_const_iterator
+    {
+        using iterator_concept = std::contiguous_iterator_tag;
+        using iterator_category = std::random_access_iterator_tag;
+        using value_type = T;
+        using difference_type = ptrdiff_t;
+        using pointer = const T*;
+        using reference = const T&;
+
+        small_vector_const_iterator() = default;
+
+#if _ITERATOR_DEBUG_LEVEL == 0
+        constexpr small_vector_const_iterator(pointer ptr) :
+            _ptr{ ptr }
+        {
+        }
+#else // _ITERATOR_DEBUG_LEVEL == 0
+        constexpr small_vector_const_iterator(pointer ptr, pointer begin, pointer end) :
+            _ptr{ ptr },
+            _begin{ begin },
+            _end{ end }
+        {
+        }
+#endif // _ITERATOR_DEBUG_LEVEL == 0
+
+        [[nodiscard]] constexpr reference operator*() const noexcept
+        {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+            _STL_VERIFY(_begin, "cannot dereference value-initialized iterator");
+            _STL_VERIFY(_ptr < _end, "cannot dereference end iterator");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+            return *_ptr;
+        }
+
+        [[nodiscard]] constexpr pointer operator->() const noexcept
+        {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+            _STL_VERIFY(_begin, "cannot dereference value-initialized iterator");
+            _STL_VERIFY(_ptr < _end, "cannot dereference end iterator");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+            return _ptr;
+        }
+
+        constexpr small_vector_const_iterator& operator++() noexcept
+        {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+            _STL_VERIFY(_begin, "cannot increment value-initialized iterator");
+            _STL_VERIFY(_ptr < _end, "cannot increment iterator past end");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+            ++_ptr;
+            return *this;
+        }
+
+        constexpr small_vector_const_iterator operator++(int) noexcept
+        {
+            small_vector_const_iterator _Tmp{ *this };
+            ++*this;
+            return _Tmp;
+        }
+
+        constexpr small_vector_const_iterator& operator--() noexcept
+        {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+            _STL_VERIFY(_begin, "cannot decrement value-initialized iterator");
+            _STL_VERIFY(_begin < _ptr, "cannot decrement iterator before begin");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+            --_ptr;
+            return *this;
+        }
+
+        constexpr small_vector_const_iterator operator--(int) noexcept
+        {
+            small_vector_const_iterator _Tmp{ *this };
+            --*this;
+            return _Tmp;
+        }
+
+        constexpr void _Verify_offset([[maybe_unused]] const difference_type _Off) const noexcept

```

Comment:
This is just a nit, but I thought we were encouraged not to use `_Upper` names in new code because they're reserved by the standard (see #721). I'm not a language lawyer, though, so I may be misinterpreting that issue.

Reply:
That member function is used internally in Microsoft's STL for `_Offset_verifiable_v` and others in `<xutility>`. It's a bit scary to depend on internals like that but I figured it's easier to just copy/paste the `std::span<>::iterator` class wholesale and get the iterators to work just like a `std::span` would. GSL depends on it as well btw: https://github.com/microsoft/GSL/blob/330583f47800c60cf001239550d291d16274756a/include/gsl/span#L292-L335

(On that note, I just realized I forgot to copy over the `_Verify_range` member.)

Reply:
I've pushed a commit which at least fixes all the other occurrences of `_Ugly` names in that file.
Also, today I realized we can use C++20 concepts. It works great!